--- src/Molecule.C	2014-08-29 17:03:06.000000000 -0400
+++ src/Molecule.C	2018-07-18 19:54:03.139813842 -0400
@@ -1215,7 +1215,8 @@
       atoms[atom_number-1].status |= LonepairAtom;
     } else if (atoms[atom_number-1].mass < 1.0) {
       atoms[atom_number-1].status |= DrudeAtom;
-    } else if (atoms[atom_number-1].mass <=3.5) {
+    } else if ((atoms[atom_number-1].mass <=4.0) &&  //2/17: VO Changing H mass cutoff to 4 so that we can use heavy hydrogens ; Note that mass(He) = 4.00262>4 amu.
+               (atomNames[atom_number-1].atomname[0] != 'C')) { // with 4amu H methyl C arbons come out at 3.5 amu
       atoms[atom_number-1].status |= HydrogenAtom;
     } else if ((atomNames[atom_number-1].atomname[0] == 'O') && 
          (atoms[atom_number-1].mass >= 14.0) && 
@@ -2574,7 +2575,8 @@
             atoms[i].status |= LonepairAtom;
         }else if(atoms[i].mass < 1.0) {
             atoms[i].status |= DrudeAtom;
-        }else if(atoms[i].mass <= 3.5) {
+        }else if((atoms[i].mass <= 4.0) &&
+                 (atomNames[i].atomname[0] != 'C')) {
             atoms[i].status |= HydrogenAtom;
         }else if((atomNames[i].atomname[0] == 'O') &&
                  (atoms[i].mass>=14.0) && (atoms[i].mass<=18.0)){
@@ -4713,7 +4715,8 @@
       fAtom->status |= LonepairAtom;
   } else if (thisAtomMass < 1.0) {
       fAtom->status |= DrudeAtom;
-  } else if (thisAtomMass <= 3.5) {
+  } else if ((thisAtomMass <= 4.0) &&
+      (thisAtomName[0]!='C')) {
       fAtom->status = HydrogenAtom;
   } else if (thisAtomName[0]=='O' &&
              (thisAtomMass >= 14.0) && (thisAtomMass <= 18.0)) {
@@ -9497,7 +9500,8 @@
       atoms[i].status |= LonepairAtom;
     } else if (atoms[i].mass < 1.0) {
       atoms[i].status |= DrudeAtom;
-    } else if (atoms[i].mass <=3.5) {
+    } else if ((atoms[i].mass <=4.0) &&
+      (atomNames[i].atomname[0] != 'C')) {
       atoms[i].status |= HydrogenAtom;
     } else if ((atomNames[i].atomname[0] == 'O') && 
          (atoms[i].mass >= 14.0) && 
