--- Make.depends	2014-11-15 11:30:50.000000000 -0500
+++ Make.depends	2017-02-27 14:54:07.000000000 -0500
@@ -3880,6 +3880,23 @@
 	src/fitrms.h \
 	src/Debug.h
 	$(CXX) $(CXXFLAGS) $(COPTO)obj/GlobalMasterTMD.o $(COPTC) src/GlobalMasterTMD.C
+obj/GlobalMasterDY.o: \
+	src/GlobalMasterDY.C \
+	src/InfoStream.h \
+	src/NamdTypes.h \
+	src/common.h \
+	src/Vector.h \
+	src/ResizeArray.h \
+	src/ResizeArrayRaw.h \
+	src/GlobalMaster.h \
+	src/SimParameters.h \
+	src/Lattice.h \
+	src/Tensor.h \
+	src/MGridforceParams.h \
+	src/strlib.h \
+	src/MStream.h \
+	src/GlobalMasterDY.h
+	$(CXX) $(CXXFLAGS) $(COPTO)obj/GlobalMasterDY.o $(COPTC) src/GlobalMasterDY.C
 obj/Matrix4Symmetry.o: \
 	src/Matrix4Symmetry.C \
 	src/Matrix4Symmetry.h \
--- Makefile	2014-11-15 11:46:41.000000000 -0500
+++ Makefile	2018-04-25 13:18:09.709792112 -0400
@@ -68,7 +68,7 @@
 #EXTRADEFINES=-DREMOVE_PROXYDATAMSG_EXTRACOPY -DREMOVE_PROXYRESULTMSG_EXTRACOPY
 EXTRADEFINES=-DREMOVE_PROXYRESULTMSG_EXTRACOPY -DNODEAWARE_PROXY_SPANNINGTREE -DUSE_NODEPATCHMGR
 EXTRAINCS=
-EXTRALINKLIBS=
+EXTRALINKLIBS+=-L$(DYNAMO_LIBDIR) -ldynamo
 # to compile namd using PAPI counters to measure flops and modify include and library path
 # correspondingly
 #EXTRADEFINES=-DREMOVE_PROXYRESULTMSG_EXTRACOPY -DMEASURE_NAMD_WITH_PAPI
@@ -194,6 +194,7 @@
 	$(DSTDIR)/GlobalMasterTcl.o \
 	$(DSTDIR)/GlobalMasterSMD.o \
 	$(DSTDIR)/GlobalMasterTMD.o \
+	$(DSTDIR)/GlobalMasterDY.o \
	$(DSTDIR)/Matrix4Symmetry.o \
	$(DSTDIR)/GlobalMasterSymmetry.o \
 	$(DSTDIR)/GlobalMasterFreeEnergy.o \
--- src/ComputeMgr.C	2013-12-27 14:59:49.000000000 -0500
+++ src/ComputeMgr.C	2018-04-25 15:52:14.226220983 -0400
@@ -8,6 +8,7 @@
 #include "ProcessorPrivate.h"
 
 //#define DEBUGM
+
 #define MIN_DEBUG_LEVEL 1
 #include "Debug.h"
 
@@ -83,6 +84,8 @@
 #include "GlobalMasterIMD.h"
 #include "GlobalMasterTcl.h"
 #include "GlobalMasterSMD.h"
+#include <stdbool.h> // VO Dynamo
+#include "GlobalMasterDY.h" // VO Dynamo
 #include "GlobalMasterTMD.h"
 #include "GlobalMasterSymmetry.h"
 #include "GlobalMasterEasy.h"
@@ -94,6 +97,7 @@
 
 ComputeMgr::ComputeMgr()
 {
+    DebugM(1,"ComputeMgr constructor called\n");
     CkpvAccess(BOCclass_group).computeMgr = thisgroup;
     computeGlobalObject = 0;
     computeGlobalResultsMsgSeq = -1;
@@ -120,7 +124,9 @@
 
 ComputeMgr::~ComputeMgr(void)
 {
+    DebugM(1,"ComputeMgr destructor called\n");
     delete computeNonbondedWorkArrays;
+//    if (masterServerObject) delete masterServerObject; // make sure to call client destructors
 }
 
 void ComputeMgr::updateComputes(int ep, CkGroupID chareID)
@@ -850,6 +856,9 @@
                                     simParams->firstTimestep, simParams->SMDFile,
                                     node->molecule->numAtoms)
             );
+// VO Dynamo v
+        if (simParams->DYPluginActive) masterServerObject->addClient(new GlobalMasterDY());
+// VO Dynamo ^
             
         if (simParams->symmetryOn && 
           (simParams->firstTimestep < simParams->symmetryLastStep || 
--- src/SimParameters.C	2014-09-05 16:15:44.000000000 -0400
+++ src/SimParameters.C	2017-02-28 11:45:50.789006340 -0500
@@ -1530,6 +1530,11 @@
    opts.range("SMDOutputFreq", POSITIVE);
    
    //****** END SMD constraints changes 
+   //****** VO: BEGIN Dynamo plugin parameters
+   opts.optionalB("main", "DynamoPlugin", "Plugin to use restraints from dynamo suite", &DYPluginActive, FALSE);
+   opts.require("DynamoPlugin", "DynamoConfigFile","Configuration file for dynamo plugin", DYConfigFileName);
+   opts.optional("DynamoPlugin", "DynamoLogFile","Log file for dynamo plugin", DYLogFileName);
+   //****** VO: END Dynamo plugin parameters
 
    //****** BEGIN tabulated energies section
    opts.optionalB("main", "tabulatedEnergies", "Do we get energies from a table?", &tabulatedEnergies, FALSE);
@@ -4383,7 +4388,11 @@
    }
    
    //****** END SMD constraints changes 
-
+   //****** VO Dynamo Plugin
+   if (DYPluginActive){
+    iout <<iINFO<<"DYNAMO PLUGIN ACTIVE\n";
+   }
+   //****** VO END Dynamo Plugin
    if (TMDOn) {
      iout << iINFO << "TMD ACTIVE BETWEEN STEPS " << TMDFirstStep 
           << " and " << TMDLastStep << "\n";
@@ -4577,7 +4586,7 @@
 
    globalForcesOn = ( tclForcesOn || freeEnergyOn || miscForcesOn ||
                       (IMDon && ! (IMDignore || IMDignoreForces)) || SMDOn || TMDOn || 
-                      colvarsOn || symmetryOn || qmForcesOn );
+                      colvarsOn || symmetryOn || qmForcesOn || DYPluginActive );
 
 
    if (tclForcesOn)
--- src/SimParameters.h	2014-05-09 17:39:06.000000000 -0400
+++ src/SimParameters.h	2017-02-27 14:34:45.694219690 -0500
@@ -344,6 +344,12 @@
  	char SMDFile[128];		//  File for SMD information
         int SMDOutputFreq;              //  Output frequency for SMD constr.
         //****** END SMD constraints changes 
+
+        //****** VO BEGIN Dynamo Plugin changes
+        Bool DYPluginActive;
+        char DYConfigFileName[128];
+        char DYLogFileName[128];
+        //****** VO END Dynamo Plugin changes
         
   //****** BEGIN tabulated energy section
   Bool tabulatedEnergies;
--- /dev/null	2018-04-05 13:57:26.716306436 -0400
+++ src/GlobalMasterDY.C	2018-04-25 17:04:39.384308456 -0400
@@ -0,0 +1,237 @@
+#ifndef __BINDC
+#define __BINDC
+#endif
+
+#include "InfoStream.h"
+#include "NamdTypes.h"
+#include "Node.h"
+#include "Molecule.h"
+#include "Lattice.h"
+#include "GlobalMaster.h"
+#include "SimParameters.h"
+//#include "ReductionMgr.h"
+#include <stdbool.h>
+#include "GlobalMasterDY.h"
+//#define DEBUGM
+#include "Debug.h"
+
+GlobalMasterDY::GlobalMasterDY() {
+   __CCHAR *inputfile = NULL, *logfile = NULL;
+   __CINT ilen=0, llen=0, ierr=0;
+   const __CCHAR* deflogfile = "dynamo.log" ;
+   params = Node::Object()->simParameters;
+   molecule=Node::Object()->molecule ;
+   __CINT natoms=molecule->numAtoms;
+//
+   initialized=0;
+   sm_energy=0.;
+   CkPrintf("# DYNAMO PLUGIN: Initializing ...\n");
+//
+// find input file
+//
+   inputfile = params->DYConfigFileName ;
+   if (!inputfile) {
+    CkPrintf("# DYNAMO PLUGIN: input file not specified (syntax : DynamoConfigFile <inputfile>)\n");
+    CkExit();
+   } else {
+    CkPrintf("# DYNAMO PLUGIN: input file is '"); 
+    CkPrintf(inputfile);
+    CkPrintf("'\n");
+   }
+//
+// find log file
+//
+   logfile = params->DYLogFileName ;
+   if (!logfile || strlen(logfile)<1) {
+    CkPrintf("# DYNAMO PLUGIN: log file not specified (syntax : DynamoLogFile <logfile>), will write to '");
+    logfile=(__CCHAR *) deflogfile;
+    CkPrintf(logfile);
+    CkPrintf("'\n");
+   } else {
+     CkPrintf("# DYNAMO PLUGIN: log file is '"); 
+     CkPrintf(logfile);
+     CkPrintf("'\n");
+   }
+//
+   __CFLOAT *mass=NULL, *charge=NULL;
+//
+   mass =   (__CFLOAT *) calloc(molecule->numAtoms, sizeof(__CFLOAT));
+   charge = (__CFLOAT *) calloc(molecule->numAtoms, sizeof(__CFLOAT));
+//
+   for (int i=0; i<molecule->numAtoms; i++){
+    mass[i]=molecule->atommass(i);
+    charge[i]=molecule->atomcharge(i);
+   }
+//
+   ilen=strlen(inputfile);
+   llen=strlen(logfile);
+// PBC
+//   CkPrintf("# DYNAMO PLUGIN: checking if simulation uses periodic boundary conditions\n"); 
+   Lattice &lattice = params->lattice;
+   usesPeriodic = ( lattice.a_p() && lattice.b_p() && lattice.c_p() ) ;
+   if (usesPeriodic) {
+    CkPrintf("# DYNAMO PLUGIN: Simulation uses periodic boundary conditions\n"); 
+    Vector const a = lattice.a() ;
+    Vector const b = lattice.b() ;
+    Vector const c = lattice.c() ;
+    Position const o = lattice.origin() ;
+    box[0]=a.x;
+    box[1]=a.y;
+    box[2]=a.z;
+    box[3]=b.x;
+    box[4]=b.y;
+    box[5]=b.z;
+    box[6]=c.x;
+    box[7]=c.y;
+    box[8]=c.z;
+    box[9]=o.x;
+    box[10]=o.y;
+    box[11]=o.z;
+   } else {
+    for ( int i=0 ; i < 12 ; i++ ) { box[i]=0.0 ; }
+   }
+// initialize dynamo
+   ierr=master_init_plugin(natoms, 0, mass, NULL, charge, NULL, inputfile, ilen, logfile, llen, &atomlist, usesPeriodic, box, NULL);
+// mass and charge no longer needed
+   free(mass);
+   free(charge);
+//
+   int atomid;
+   if (atomlist!=NULL) { // atom indices provided; add them
+    // first element gives list size:
+    for ( int l = 0 ; l++ < atomlist[0]  ; ) { // first value in atomlist is the list length
+     atomid = atomlist[l] - 1 ; // subtract one because atom indices are offset from 0 in NAMD, but from 1 in the DY plugin
+     if ( atomid >= 0 && atomid < molecule->numAtoms ) {
+      DebugM(1,"Adding atom "<<atomid<<"\n");
+      modifyRequestedAtoms().add(atomid); // note: atomlist will be deallocated in Fortran
+     } else {
+      CkPrintf("# DYNAMO PLUGIN: Atom ID ");
+      CkPrintf("%d", atomid);
+      CkPrintf("is out of range\n");
+      CkExit();
+     } // atomid valid
+    } // over list
+   } else { // no indices provided, assume that initialization is deferred until dynamics; request all coordinates
+    for ( atomid = 0 ; atomid < molecule->numAtoms ; atomid++) {
+      DebugM(1,"Adding atom "<<atomid<<"\n");
+      modifyRequestedAtoms().add(atomid);
+    }
+   } // atomlist
+   r  = (__CFLOAT *) calloc(3*molecule->numAtoms, sizeof(__CFLOAT));
+   fr = (__CFLOAT *) calloc(3*molecule->numAtoms, sizeof(__CFLOAT));
+//
+//   iteration = params->firstTimestep; // not sure we need to keep a local counter
+//   reduction = ReductionMgr::Object()->willSubmit(REDUCTIONS_BASIC);
+} // GlobalMasterDY
+
+
+GlobalMasterDY::~GlobalMasterDY() {
+ params=NULL;
+ destroy();
+} // ~GlobalMasterDY
+
+
+void GlobalMasterDY::calculate() {
+ // load coordinates
+  int atomid;
+  AtomIDList::const_iterator a_i = getAtomIdBegin();
+  AtomIDList::const_iterator a_e = getAtomIdEnd();
+  PositionList::const_iterator p_i = getAtomPositionBegin();
+  Vector p ;
+  for ( ; a_i != a_e; ++a_i, ++p_i ) { // this loop is only over atoms that have been requested
+    atomid = (*a_i);
+#ifdef DEBUGM
+    CkPrintf("%d\n",atomid);
+#endif
+    atomid*=3; // index into coordinate array
+    p = *p_i;
+    r[atomid++] = p.x; // a_i dereferences to atomid ; p_i dereferences to position (instance of Vector)
+    r[atomid++] = p.y; // however, in principle, access can be random (at the very least, nonconsecutive)
+    r[atomid]   = p.z;
+  }
+//
+#ifdef DEBUGM
+// check that forces are received correctly
+  for (int i=0;i<3*molecule->numAtoms;i++){
+   CkPrintf("%15.10f %15.10f\n",r[i],fr[i]);
+  }
+#endif
+  if (usesPeriodic) {
+   Vector const a = lattice->a() ;
+   Vector const b = lattice->b() ;
+   Vector const c = lattice->c() ;
+   Position const o = lattice->origin() ;
+   box[0]=a.x;
+   box[1]=a.y;
+   box[2]=a.z;
+   box[3]=b.x;
+   box[4]=b.y;
+   box[5]=b.z;
+   box[6]=c.x;
+   box[7]=c.y;
+   box[8]=c.z;
+   box[9]=o.x;
+   box[10]=o.y;
+   box[11]=o.z;
+  }
+  // apply forces
+  modifyAppliedForces().resize(0);
+  modifyForcedAtoms().resize(0);
+  //
+  if (atomlist!=NULL) { // atom indices already provided; compute and add forces
+    //call sm
+    __CINT ierr=master_dyna_plugin(step, 0, r, NULL, fr, NULL, &sm_energy, NULL, &atomlist, usesPeriodic, box, NULL);
+    // first element gives list size:
+   for ( int l = 0 ; l++ < atomlist[0]  ; ) { // first value in atomlist is the list length
+     atomid = atomlist[l] - 1; // subtract one because atom indices are offset from 0 in NAMD, but from 1 in the DY plugin
+     modifyForcedAtoms().add(atomid);
+     atomid*=3 ; // index into force array
+     p.x=fr[atomid++];
+     p.y=fr[atomid++];
+     p.z=fr[atomid];
+     DebugM(1,"Adding force "<< p << " to atom "<<atomid<<"\n");
+     modifyAppliedForces().add(p);
+   } // for
+  } else { // atomlist is NULL, but we expect it to be populated by the plugin at the first dynamics step
+    //call sm
+    __CINT ierr=master_dyna_plugin(step, 0, r, NULL, fr, NULL, &sm_energy, NULL, &atomlist, usesPeriodic, box, NULL);
+    //
+    if (atomlist!=NULL) { // atom indices provided; add them
+     modifyRequestedAtoms().resize(0);
+    // first element gives list size:
+     for ( int l = 0 ; l++ < atomlist[0]  ; ) { // first value in atomlist is the list length
+      atomid = atomlist[l] - 1 ; // subtract one because atom indices are offset from 0 in NAMD, but from 1 in the DY plugin
+      if ( atomid >= 0 && atomid < molecule->numAtoms ) {
+       DebugM(1,"Adding atom "<<atomid<<"\n");
+       modifyRequestedAtoms().add(atomid); // note: atomlist will be deallocated in Fortran
+       // now add forces :
+       modifyForcedAtoms().add(atomid);
+       atomid*=3 ; // index into force array
+       p.x=fr[atomid++];
+       p.y=fr[atomid++];
+       p.z=fr[atomid];
+       DebugM(1,"Adding force "<< p << " to atom "<<atomid<<"\n");
+       modifyAppliedForces().add(p);
+      } else {
+       CkPrintf("# DYNAMO PLUGIN: Atom ID ");
+       CkPrintf("%d", atomid);
+       CkPrintf("is out of range\n");
+       CkExit();
+      } // atomid valid
+     } // over list
+   } else { // atom list still undefined
+    CkPrintf("# DYNAMO PLUGIN: No restraint atoms found (nothing to do)\n");
+    CkExit(); 
+   } // atomlist
+  }
+  // destructor is not always called, so finalize after run length exceeded
+  if (step>params->N) destroy(); // this is problematic for multiple runs per script, e.g. minimize, then run MD
+} // calculate
+
+void GlobalMasterDY::destroy(){
+  CkPrintf("# DYNAMO PLUGIN: Finalizing...\n");
+  master_done_plugin();
+// Keep in case there are multiple runs
+//  if (r) free(r);
+//  if (fr) free(fr);
+}
--- /dev/null	2018-04-05 13:57:26.716306436 -0400
+++ src/GlobalMasterDY.h	2018-04-25 17:19:21.638272676 -0400
@@ -0,0 +1,37 @@
+/* C prototypes */
+// C, not Fortran!
+extern "C" int master_init_plugin(const int, const _Bool, const double *, const float *, const double *, const float *, const char *, const int, const char *, const int, int **, const _Bool, const double *, const float *);
+extern "C" int master_dyna_plugin(const long int, const _Bool, const double *, const float *, double *, float *, double *, float *, int **, const _Bool, const double *, const float * );
+extern "C" void master_done_plugin();
+class Molecule;
+class SubmitReduction;
+class GlobalMasterDY : public GlobalMaster {
+ public:
+  GlobalMasterDY();
+  ~GlobalMasterDY();
+ private:
+  Bool initialized;
+  void calculate();
+  void destroy();
+  Molecule *molecule;
+  SimParameters * params;
+  long int iteration; // not sure we need to keep a local counter
+//  SubmitReduction *reduction;
+  double sm_energy;
+  int *atomlist=NULL; // to maintain a list of atoms that are needed by plugin; memory managed by FORTRAN side
+  double *r=NULL, *fr=NULL ; //positions and forces
+  double box[12] ; // three unit cell vectors followed by cell origin
+  _Bool usesPeriodic ;// whether the simulation uses PBC
+};
+#if defined(__BINDC)
+#define __FCFLOAT c_double 
+#define __CFLOAT double 
+#define __CFLOATFMT "%lf"
+#define __FCINT c_int
+#define __CINT int
+#define __CINTFMT "%d"
+#define __FCBOOL c_bool
+#define __CBOOL _Bool
+#define __FCCHAR c_char
+#define __CCHAR char
+#endif
