--- ../gnu/source/dynamc/dynamc.src	2020-07-31 12:37:18.450134312 -0400
+++ ./source/dynamc/dynamc.F90	2020-08-17 11:24:54.013592355 -0400
@@ -804,12 +804,12 @@
 #if KEY_STRINGM==1 /*  VO stringm */
 !================ call string method routines ========================
       if (smcv_on) call smcv_main(x,y,z,xcomp,ycomp,zcomp,&
-                                  amass(1:natom),dx,dy,dz,istart-1)
+                                  amass(1:natom),dx,dy,dz,int8(istart-1))
       if (ftsm_on) call ftsm_main(x(1:natom),y(1:natom),z(1:natom),             &
      &                            xcomp(1:natom),ycomp(1:natom),zcomp(1:natom), &
      &                            dx(1:natom),dy(1:natom),dz(1:natom),          &
      &                            amass(1:natom),                               &
-     &                            istart-1,wmain(1:natom),bnbnd,bimag)
+     &                            int8(istart-1),wmain(1:natom),bnbnd,bimag)
 !=====================================================================
 #endif
         call timer_start(T_dcntrl)
@@ -995,12 +995,12 @@
 #if KEY_STRINGM==1 /*  VO stringm */
 !================ call string method routines ========================
       if (smcv_on) call smcv_main(x,y,z,xcomp,ycomp,zcomp,&
-                                  amass(1:natom),dx,dy,dz,istart-1)
+                                  amass(1:natom),dx,dy,dz,int8(istart-1))
       if (ftsm_on) call ftsm_main(x(1:natom),y(1:natom),z(1:natom),             &
      &                            xcomp(1:natom),ycomp(1:natom),zcomp(1:natom), &
      &                            dx(1:natom),dy(1:natom),dz(1:natom),          &
      &                            amass(1:natom),                               &
-     &                            istart-1,wmain(1:natom),bnbnd,bimag)
+     &                            int8(istart-1),wmain(1:natom),bnbnd,bimag)
 !=====================================================================
 #endif
      call timer_start(T_dcntrl)
@@ -1965,7 +1965,7 @@
      !
      !  check replica cell
         if (.not.smcv_voronoi_compute(xcomp,ycomp,zcomp,&
-            amass(1:natom),istep)) then ! check V. cell
+            amass(1:natom),int8(istep))) then ! check V. cell
          if (voronoi_wrong_cell) then
            if (ME_LOCAL.eq.0) &
            write(outu,770) ME_STRNG, istep
@@ -2065,7 +2065,7 @@
          endif ! map =/= -1
         endif  ! compute_whereami
      !
-        if (.not.ftsm_voronoi_check(xcomp,ycomp,zcomp,istep)) then ! still in the current cell?
+        if (.not.ftsm_voronoi_check(xcomp,ycomp,zcomp,int8(istep))) then ! still in the current cell?
          if (voronoi_wrong_cell) then
            if (ME_LOCAL.eq.0) &
            write(outu,771) ME_STRNG, istep
@@ -2774,12 +2774,12 @@
 #if KEY_STRINGM==1 /*  VO stringm */
      !=========================================
      if (smcv_on) call smcv_main(x,y,z,xcomp,ycomp,zcomp,&
-     &                            amass(1:natom),dx,dy,dz,istep)
+     &                            amass(1:natom),dx,dy,dz,int8(istep))
      if (ftsm_on) call ftsm_main(x(1:natom),y(1:natom),z(1:natom),&
      &                            xcomp(1:natom),ycomp(1:natom),zcomp(1:natom),&
      &                            dx(1:natom),dy(1:natom),dz(1:natom),&
      &                            amass(1:natom),&
-     &                            istep,wmain(1:natom),bnbnd,bimag)
+     &                            int8(istep),wmain(1:natom),bnbnd,bimag)
      !=========================================
 #endif /* stringm */
      !
