diff --git a/source/ensemble/ensemble.F90 b/source/ensemble/ensemble.F90
index 617990a..0acdda1 100644
--- source/ensemble/ensemble.F90
+++ source/ensemble/ensemble.F90
@@ -62,25 +62,6 @@ subroutine ensini
   ensexpu = -1
 
   if (iolev > 0 .and. ensemble_verbose) then
-#if KEY_STRINGM==1
-#if KEY_ABPO==1
-     write(outu,'(a)')  &
-          ' ENSINI> ALL REPLICAS PRESENTLY USE ONE TEMPERATURE', &
-          ' ENSINI> USE "ENSEmble EXCHange"', &
-          ' ENSINI> TO SET UP REPLICA EXCHANGE.', &
-          ' ENSINI> USE ENSEmble STRIng [...]', &
-          ' ENSINI> TO USE THE STRING METHOD.', &
-          ' ENSINI> USE ENSEmble ABPO [...]', &
-          ' ENSINI> TO USE THE ABPO METHOD.'
-#else /**/
-     write(outu,'(a)')  &
-          ' ENSINI> ALL REPLICAS PRESENTLY USE ONE TEMPERATURE', &
-          ' ENSINI> USE "ENSEmble EXCHange"', &
-          ' ENSINI> TO SET UP REPLICA EXCHANGE.', &
-          ' ENSINI> USE ENSEmble STRIng [...]', &
-          ' ENSINI> TO USE THE STRING METHOD.'
-#endif 
-#else /**/
 #if KEY_ABPO==1
      write(outu,'(a)')  &
           ' ENSINI> ALL REPLICAS PRESENTLY USE ONE TEMPERATURE', &
@@ -93,7 +74,6 @@ subroutine ensini
           ' ENSINI> ALL REPLICAS PRESENTLY USE ONE TEMPERATURE', &
           ' ENSINI> USE "ENSEmble EXCHange"', &
           ' ENSINI> TO SET UP REPLICA EXCHANGE.'
-#endif 
 #endif 
   endif
   return
@@ -261,10 +241,6 @@ subroutine enscmd(comlyn,comlen)
      call enswri(comlyn,comlen)
    else if (indxa(comlyn,comlen,'EVB') > 0) then
       call evbsetup(comlyn, comlen)
-#if KEY_STRINGM==1
-  else if (indxa(comlyn,comlen,'STRI')  >  0) then
-     CALL parse_string_commands(comlyn,comlen) ! parse string commands
-#endif 
 #if KEY_ABPO==1
   else if (indxa(comlyn,comlen,'ABPO')  >  0) then
      call abpo_setup(comlyn,comlen) ! parse abpo commands
