diff --git a/source/rxncor/travel.F90 b/source/rxncor/travel.F90
index c344895..3be0190 100644
--- source/rxncor/travel.F90
+++ source/rxncor/travel.F90
@@ -407,26 +407,26 @@ contains
          CALL ADDPRL(SERIAL(IDX),SERIAL(IDXNEW), NEIBR1,NEIBR2,NEIBOR)
          CALL ADDPRL(SERIAL(IDXNEW),SERIAL(IDXFOL),NEIBR1,NEIBR2,NEIBOR)
 
-         I = IDX
-         CALL SGSCAN( L,I,IDNEXT,PTENE,NMAXI,LSCAN1,LSCAN2,LSCAN3, &
+!         I = IDX ! no need because sgscan does not modify IDX
+         CALL SGSCAN( L,IDX,IDNEXT,PTENE,NMAXI,LSCAN1,LSCAN2,LSCAN3, &
               LENSEG,STPMAX,NEXMIN,ENEMAX,QUP,QDOWN,PREMIN, &
               SRTENE,SRTIDX,SEGSTP,PTGRAD, &
               XBAS,YBAS,ZBAS, NATOM,VERBEU )
          CALL UPDNEW( I,LPAKED,SRTENE,SRTIDX,NMAXI,STPMAX, &
               NLINMN,PTGRAD,SADGRD,SADMIN,ENEMAX )
 
-         I = IDXNEW
-         CALL SGSCAN( L,I,IDNEXT,PTENE,NMAXI,LSCAN1,LSCAN2,LSCAN3, &
+!         I = IDXNEW
+         CALL SGSCAN( L,IDXNEW,IDNEXT,PTENE,NMAXI,LSCAN1,LSCAN2,LSCAN3, &
               LENSEG,STPMAX,NEXMIN,ENEMAX,QUP,QDOWN,PREMIN, &
               SRTENE,SRTIDX,SEGSTP,PTGRAD, &
               XBAS,YBAS,ZBAS, NATOM,VERBEU )
          CALL UPDNEW( I,LPAKED,SRTENE,SRTIDX,NMAXI,STPMAX, &
               NLINMN,PTGRAD,SADGRD,SADMIN,ENEMAX )
 
-         I = IDXFOL
-         CALL UPDEND( I,LPAKED,SRTENE,SRTIDX,NMAXI,STPMAX,QUP,QDOWN, &
+!         I = IDXFOL
+         CALL UPDEND( IDXFOL,LPAKED,SRTENE,SRTIDX,NMAXI,STPMAX,QUP,QDOWN, &
               NLINMN,PTGRAD,SADGRD,SADMIN,PTENE,ENEMAX )
-
+! VO : a question that remains is whether there is a change that mods to global i were intentional ...
          NEWCYC = IDXNEW
       ENDIF
 
diff --git a/source/rxncor/travel2.F90 b/source/rxncor/travel2.F90
index 220fdf2..55a321f 100644
--- source/rxncor/travel2.F90
+++ source/rxncor/travel2.F90
@@ -193,7 +193,8 @@ contains
 
 
     LOGICAL   LPAKED,QUP(*),QDOWN(*)
-    INTEGER   I, SRTIDX(*),NMAXI,STPMAX(*),NLINMN(*),SADMIN
+    integer, intent(in) :: i ! make sure not to modify VO
+    INTEGER   SRTIDX(*),NMAXI,STPMAX(*),NLINMN(*),SADMIN
     real(chm_real)    SRTENE(*),PTGRAD(*),SADGRD,PTENE(*),ENEMAX(*)
 
     INTEGER, PARAMETER :: BIGINT=999999
@@ -3231,7 +3232,8 @@ contains
 
     real(chm_real),allocatable,dimension(:) :: IDUMX,IDUMY,IDUMZ
     LOGICAL   LSCAN1,LSCAN2,LSCAN3, QUP(*),QDOWN(*)
-    INTEGER   LLL,I,IDNEXT(*),NMAXI, STPMAX(*),NEXMIN(*),PREMIN(*)
+    integer, intent(in) :: I ! VO make sure i does not change
+    INTEGER   LLL,IDNEXT(*),NMAXI, STPMAX(*),NEXMIN(*),PREMIN(*)
     INTEGER   SRTIDX(*), NATOM,VERBEU
     type(chm_array)   XBAS(:),YBAS(:),ZBAS(:)
     real(chm_real)    PTENE(*),LENSEG(*),ENEMAX(*),SRTENE(*),SEGSTP(*)
